### PR TITLE
[VPC-3359] Implemented vpc peering methods

### PR DESCRIPTION
**Pull request by guptado**
VPC Peering is a closed beta feature and not available for public use.

[Testing doc ](https://docs.google.com/document/d/1kE451w61E_zhOZ04bTTkvEBJiLSQZWlHVtupRs8oalg/edit)

[VPC-3359]: https://do-internal.atlassian.net/browse/VPC-3359?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: commands/vpc_peerings_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		vpcID := "vpc-01"
+		tm.vpcs.EXPECT().ListVPCPeeringsByVPCID(vpcID).Return(testPeeringList, nil)
+
+		config.Doit.Set(config.NS, doctl.ArgVPCPeeringVPCID, vpcID)
+		err := RunVPCPeeringList(config)
+		assert.NoError(t, err)
+	})
+}
+
+func TestVPCPeeringCreate(t *testing.T) {
```
Comment (lfundaro-do) [inline commands/vpc_peerings_test.go:71]: should we add additional test cases to validate missing vpc ids ?

File: do/vpcs.go [hunk 2]
```diff
@@ -119,3 +133,93 @@ func (v *vpcsService) Delete(vpcUUID string) error {
...
+func (v *vpcsService) GetPeering(peeringID string) (*VPCPeering, error) {
+	peering, _, err := v.client.VPCs.GetVPCPeering(context.TODO(), peeringID)
+	if err != nil {
+		return nil, err
+	}
+	return &VPCPeering{VPCPeering: peering}, nil
+}
+
+func (v *vpcsService) ListVPCPeerings() (VPCPeerings, error) {
+
```
Comment (lfundaro-do) [inline do/vpcs.go:146]:
```suggestion
```

File: do/vpcs.go [hunk 3]
```diff
@@ -119,3 +133,93 @@ func (v *vpcsService) Delete(vpcUUID string) error {
...
+func (v *vpcsService) CreateVPCPeering(req *godo.VPCPeeringCreateRequest) (*VPCPeering, error) {
+	peering, _, err := v.client.VPCs.CreateVPCPeering(context.TODO(), req)
+	if err != nil {
+		return nil, err
+	}
+	return &VPCPeering{VPCPeering: peering}, nil
+}
+
+func (v *vpcsService) ListVPCPeeringsByVPCID(vpcID string) (VPCPeerings, error) {
+
```
Comment (lfundaro-do) [inline do/vpcs.go:198]:
```suggestion
```

File: commands/vpc_peerings.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+	AddStringFlag(cmdPeeringCreate, doctl.ArgVPCPeeringFirstVPCID, "", "",
+		"First VPC ID")
+	AddStringFlag(cmdPeeringCreate, doctl.ArgVPCPeeringSecondVPCID, "", "",
+		"Second VPC ID")
+	cmdPeeringCreate.Example = `The following example creates a VPC Peering named ` +
+		"`" + `example-peering` + "`" +
+		` : doctl vpc-peerings create --name example-peering --first-vpc-id f81d4fae-7dec-11d0-a765-00a0c91e6bf6 --second-vpc-id 3f900b61-30d7-40d8-9711-8c5d6264b268`
+
+	cmdPeeringUpdate := CmdBuilder(cmd, RunVPCPeeringUpdate, "update <id>",
+		"Update a VPC Peering's configuration", `Updates a VPC network's configuration. You can update its name.`, Writer, aliasOpt("u"))
```
Comment (ddatta-do) [inline commands/vpc_peerings.go:59]: wondering if we should rephrase this line,  as we really cannot `Updates a VPC network's configuration` through this

File: commands/vpc_peerings.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+	peeringDetails := `
+- The VPC Peering ID
+- The VPC Peering Name
+- The First VPC network's ID
+- The Second VPC network's ID
+- The VPC Peering Status
+- The VPC Peering creation date, in ISO8601 combined date and time format
+`
+	cmdPeeringGet := CmdBuilder(cmd, RunVPCPeeringGet, "get <id>",
+		"Retrieve a VPC network", "Retrieve information about a VPC Peering, including:"+peeringDetails, Writer,
```
Comment (ddatta-do) [inline commands/vpc_peerings.go:36]:
```suggestion
		"Retrieve a VPC network", "Retrieves information about a VPC Peering, including:"+peeringDetails, Writer,
```

File: commands/displayers/vpc_peering.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+func (v *VPCPeering) JSON(out io.Writer) error {
+	return writeJSON(v.VPCPeerings, out)
+}
+
+func (v *VPCPeering) Cols() []string {
+	return []string{
+		"ID",
+		"Name",
+		"FirstVpcID",
+		"SecondVpcID",
```
[the remark this answers is not in the document]
Comment (guptado) [reply commands/displayers/vpc_peering.go:23]: afaik we are accepting vpcIDs as a slice and nothing else is mention in the docs.

File: commands/vpc_peerings.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	cmdPeeringUpdate.Example = `The following example updates the name of a VPC Peering with the ID ` +
+		"`" + `f81d4fae-7dec-11d0-a765-00a0c91e6bf6` + "`" + ` to ` + "`" + `new-name` + "`" +
+		`: doctl vpc-peerings update f81d4fae-7dec-11d0-a765-00a0c91e6bf6 --name new-name`
+
+	cmdPeeringDelete := CmdBuilder(cmd, RunVPCPeeringDelete, "delete <id>",
+		"Permanently delete a VPC Peering", `Permanently deletes the specified VPC Peering. This is irreversible.`, Writer, aliasOpt("d", "rm"))
+	AddBoolFlag(cmdPeeringDelete, doctl.ArgForce, doctl.ArgShortForce, false,
+		"Delete the VPC Peering without a confirmation prompt")
+	cmdPeeringDelete.Example = `The following example deletes the VPC Peering with the ID ` + "`" + `f81d4fae-7dec-11d0-a765-00a0c91e6bf6` + "`" +
+		`: doctl vpc-peerings delete f81d4fae-7dec-11d0-a765-00a0c91e6bf6`
```
[the remark this answers is not in the document]
Comment (guptado) [reply commands/vpc_peerings.go:71]: its [here](https://github.com/digitalocean/doctl/pull/1532/files#diff-f09c0ea5f6ad01d9f2b4eb17d2dabdb39c7c8db3f3387594f9a8de70f35b2e82R68)

File: commands/vpc_peerings.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	AddStringFlag(cmdPeeringCreate, doctl.ArgVPCPeeringSecondVPCID, "", "",
+		"Second VPC ID")
+	cmdPeeringCreate.Example = `The following example creates a VPC Peering named ` +
+		"`" + `example-peering` + "`" +
+		` : doctl vpc-peerings create --name example-peering --first-vpc-id f81d4fae-7dec-11d0-a765-00a0c91e6bf6 --second-vpc-id 3f900b61-30d7-40d8-9711-8c5d6264b268`
+
+	cmdPeeringUpdate := CmdBuilder(cmd, RunVPCPeeringUpdate, "update <id>",
+		"Update a VPC Peering's configuration", `Updates a VPC Peering's configuration. You can update its name.`, Writer, aliasOpt("u"))
+	AddStringFlag(cmdPeeringUpdate, doctl.ArgVPCPeeringName, "", "",
+		"The VPC network's name")
```
Comment (ddatta-do) [inline commands/vpc_peerings.go:61]:
```suggestion
		"The VPC Peering's name")
```

File: commands/vpc_peerings.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+	peeringDetails := `
+- The VPC Peering ID
+- The VPC Peering Name
+- The First VPC network's ID
+- The Second VPC network's ID
+- The VPC Peering Status
+- The VPC Peering creation date, in ISO8601 combined date and time format
+`
+	cmdPeeringGet := CmdBuilder(cmd, RunVPCPeeringGet, "get <id>",
+		"Retrieve a VPC network", "Retrieves information about a VPC Peering, including:"+peeringDetails, Writer,
```
Comment (ddatta-do) [inline commands/vpc_peerings.go:36]:
```suggestion
		"Retrieves a VPC Peering", "Retrieves information about a VPC Peering, including:"+peeringDetails, Writer,
```

File: commands/vpc_peerings.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+	AddStringFlag(cmdPeeringCreate, doctl.ArgVPCPeeringFirstVPCID, "", "",
+		"First VPC ID")
+	AddStringFlag(cmdPeeringCreate, doctl.ArgVPCPeeringSecondVPCID, "", "",
+		"Second VPC ID")
+	cmdPeeringCreate.Example = `The following example creates a VPC Peering named ` +
+		"`" + `example-peering` + "`" +
+		` : doctl vpc-peerings create --name example-peering --first-vpc-id f81d4fae-7dec-11d0-a765-00a0c91e6bf6 --second-vpc-id 3f900b61-30d7-40d8-9711-8c5d6264b268`
+
+	cmdPeeringUpdate := CmdBuilder(cmd, RunVPCPeeringUpdate, "update <id>",
+		"Update a VPC Peering's configuration", `Updates a VPC Peering's configuration. You can update its name.`, Writer, aliasOpt("u"))
```
Comment (ddatta-do) [inline commands/vpc_peerings.go:59]:
```suggestion
		"Update a VPC Peering's name", `Use this command to update the name of a VPC Peering`, Writer, aliasOpt("u"))
```
Is this better ?

File: commands/vpc_peerings.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+- The Second VPC network's ID
+- The VPC Peering Status
+- The VPC Peering creation date, in ISO8601 combined date and time format
+`
+	cmdPeeringGet := CmdBuilder(cmd, RunVPCPeeringGet, "get <id>",
+		"Retrieve a VPC network", "Retrieves information about a VPC Peering, including:"+peeringDetails, Writer,
+		aliasOpt("g"), displayerType(&displayers.VPCPeering{}))
+	cmdPeeringGet.Example = `The following example retrieves information about a VPC Peering with the ID ` + "`" + `f81d4fae-7dec-11d0-a765-00a0c91e6bf6` + "`" + `: doctl vpc-peerings get f81d4fae-7dec-11d0-a765-00a0c91e6bf6`
+
+	cmdPeeringList := CmdBuilder(cmd, RunVPCPeeringList, "list", "List VPC Peerings", "Retrieves a list of the VPC Peerings on your account, including the following information for each:"+peeringDetails, Writer,
```
Comment (ddatta-do) [inline commands/vpc_peerings.go:40]:
```suggestion
	cmdPeeringList := CmdBuilder(cmd, RunVPCPeeringList, "list", "List VPC Peerings", "Retrieves a list of the VPC Peerings on your account, including the following informations for each:"+peeringDetails, Writer,
```

File: commands/vpc_peerings.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+	AddStringFlag(cmdPeeringUpdate, doctl.ArgVPCPeeringName, "", "",
+		"The VPC network's name")
+	cmdPeeringUpdate.Example = `The following example updates the name of a VPC Peering with the ID ` +
+		"`" + `f81d4fae-7dec-11d0-a765-00a0c91e6bf6` + "`" + ` to ` + "`" + `new-name` + "`" +
+		`: doctl vpc-peerings update f81d4fae-7dec-11d0-a765-00a0c91e6bf6 --name new-name`
+
+	cmdPeeringDelete := CmdBuilder(cmd, RunVPCPeeringDelete, "delete <id>",
+		"Permanently delete a VPC Peering", `Permanently deletes the specified VPC Peering. This is irreversible.`, Writer, aliasOpt("d", "rm"))
+	AddBoolFlag(cmdPeeringDelete, doctl.ArgForce, doctl.ArgShortForce, false,
+		"Delete the VPC Peering without a confirmation prompt")
```
Comment (ddatta-do) [inline commands/vpc_peerings.go:69]:
```suggestion
		"Delete the VPC Peering without any confirmation prompt")
```

File: commands/vpc_peerings.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lfundaro-do) [reply commands/vpc_peerings.go:71]: yes but the example is not showing that, that's what I meant: 
`doctl vpc-peerings delete f81d4fae-7dec-11d0-a765-00a0c91e6bf6`
Vs.
`doctl vpc-peerings delete -f f81d4fae-7dec-11d0-a765-00a0c91e6bf6`

?

File: args.go [hunk 13]
```diff
@@ -468,6 +468,15 @@ const (
 	// ArgVPCIPRange is a VPC range of IP addresses in CIDR notation.
 	ArgVPCIPRange = "ip-range"
 
+	// ArgVPCPeeringName is a name of the VPC Peering.
+	ArgVPCPeeringName = "name"
+	// ArgVPCPeeringFirstVPCID is the first vpc id of the peering
+	ArgVPCPeeringFirstVPCID = "first-vpc-id"
```
Comment (viola) [inline args.go:474]: can this arg take a list of VPC IDs as in "uuid-1, uuui2"

Comment (guptado) [reply args.go:474]: yeah we can make it. I'm not sure whether it looks good.  @lfundaro-do @brunograsselli wdyt?

Comment (lfundaro-do) [reply args.go:474]: yeah, that could work, I like Viola's idea

Comment (guptado) [reply args.go]: feedback is incorporated

File: commands/vpc_peerings.go [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+		if wait {
+			notice("VPC Peering deletion is in progress, waiting for VPC Peering to be deleted")
+
+			err := waitForVPCPeering(vpcs, peeringID, "DELETED", true)
+			if err != nil {
+				return fmt.Errorf("VPC Peering couldn't be deleted : %v", err)
+			}
+		}
+
```
Comment (ddatta-do) [inline commands/vpc_peerings.go:228]: would be great to show a message saying "VPC Peering is successfully deleted" <- something like this

File: args.go [hunk 15]
```diff
@@ -468,6 +468,13 @@ const (
 	// ArgVPCIPRange is a VPC range of IP addresses in CIDR notation.
 	ArgVPCIPRange = "ip-range"
 
+	// ArgVPCPeeringName is a name of the VPC Peering.
+	ArgVPCPeeringName = "name"
+	// ArgVPCPeeringVPCIDs is the first vpc id of the peering
+	ArgVPCPeeringVPCIDs = "vpc-ids"
+	// ArgVPCPeeringVPCID is id of the VPC.
+	ArgVPCPeeringVPCID = "vpc-id"
```
[the remark this answers is not in the document]
Comment (guptado) [reply args.go:476]: yes. both have different functionalities

File: commands/vpc_peerings.go [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+		if peering.Status == errStatus {
+			return fmt.Errorf("VPC Peering (%s) entered status `%s`", peeringID, errStatus)
+		}
+
+		if peering.Status == wantStatus {
+			return nil
+		}
+
+		attempts++
+		time.Sleep(10 * time.Second)
```
[the remark this answers is not in the document]
Comment (guptado) [reply commands/vpc_peerings.go:267]: Hmm... any particular reason? I think we have 10s at other wait flag places. Just maintaining the consistency across code.

File: args.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ddatta-do) [reply args.go:476]: okay , the comments need to be fixed though
```suggestion
	// ArgVPCPeeringVPCIDs are the VPC ids of the Peering
	ArgVPCPeeringVPCIDs = "vpc-ids"
	// ArgVPCPeeringVPCID is the id of the VPC.
	ArgVPCPeeringVPCID = "vpc-id"
```

**Review by bentranter — CHANGES_REQUESTED** [review]
A few comments around documentation, and two larger design comments that I'll put here:

1. I would prefer that the new command be under just `peerings` instead of `vpc-peerings`. Having it under `vpc-peerings` makes it repetitive to type out when invoking the command – `doctl vpcs vpc-peerings --help`, vs. `doctl vpcs peerings --help` for example.
2. For the `create` subcommand, both arguments are flags. For a create command in `doctl`, the more common pattern is for the name to be a positional argument, and then other arguments to be flags. This would mean the create command is `doctl vpcs peerings create <name> --vpc-ids <ids>` instead.

(Pinging @andrewsomething and @danaelhe to see if they agree with the two design related points)

**Review by bentranter — APPROVED** [review]
Looks great, awesome work!

**Review by AungkoMCM — COMMENTED** [review]
#1534

**Pull request merged**